src/storage/mod.rs

```diff
@@ -957,25 +957,30 @@ impl<E: Engine, L: LockMgr> Storage<E, L> {
                                 ctx.get_isolation_level(),
                                 !ctx.get_not_fill_cache(),
                             );
-                            let kv_pairs: Vec<_> = snap_store
+                            let result = snap_store
                                 .batch_get(&keys, &mut statistics)
-                                .into_iter()
-                                .zip(keys)
-                                .filter(|&(ref v, ref _k)| {
-                                    !(v.is_ok() && v.as_ref().unwrap().is_none())
-                                })
-                                .map(|(v, k)| match v {
-                                    Ok(Some(x)) => Ok((k.into_raw().unwrap(), x)),
-                                    Err(e) => Err(Error::from(e)),
-                                    _ => unreachable!(),
-                                })
-                                .collect();
+                                .map_err(Error::from)
+                                .map(|v| {
+                                    let kv_pairs: Vec<_> = v
+                                        .into_iter()
+                                        .zip(keys)
+                                        .filter(|&(ref v, ref _k)| {
+                                            !(v.is_ok() && v.as_ref().unwrap().is_none())
+                                        })
+                                        .map(|(v, k)| match v {
+                                            Ok(Some(x)) => Ok((k.into_raw().unwrap(), x)),
+                                            Err(e) => Err(Error::from(e)),
+                                            _ => unreachable!(),
+                                        })
+                                        .collect();
+                                    tls_collect_key_reads(CMD, kv_pairs.len());
+                                    kv_pairs
+                                });
 
-                            tls_collect_key_reads(CMD, kv_pairs.len());
```

Comment:
Why move the line into the upper closure? I guess keep 3 `tls_collect` in one place is better.

Reply:
It is because this line needs `kv_pairs.len()` which can be get directly only in the closure. Outside the closure, it is wrapped inside a `Result<>` which becomes harder to get..

Reply:
It is because this line needs `kv_pairs.len()` which can be get directly only in the closure. Outside the closure, it is wrapped inside a `Result<>` which becomes harder to get..